src/nbl/video/IGPUCommandBuffer.cpp

```diff
@@ -682,6 +682,93 @@ bool IGPUCommandBuffer::copyImage(const IGPUImage* const srcImage, const IGPUIma
 }
 
+bool IGPUCommandBuffer::invalidShaderGroups(
+    const asset::SBufferRange<const IGPUBuffer>& raygenGroupRange, uint32_t raygenGroupStride,
+    const asset::SBufferRange<const IGPUBuffer>& missGroupsRange, uint32_t missGroupStride,
+    const asset::SBufferRange<const IGPUBuffer>& hitGroupsRange, uint32_t hitGroupStride,
+    const asset::SBufferRange<const IGPUBuffer>& callableGroupsRange, uint32_t callableGroupStride, 
+    core::bitflag<IGPURayTracingPipeline::SCreationParams::FLAGS> flags) const
+{
+
+    using PipelineFlag = IGPURayTracingPipeline::SCreationParams::FLAGS;
+    using PipelineFlags = core::bitflag<PipelineFlag>;
+
+    // https://registry.khronos.org/vulkan/specs/latest/man/html/vkCmdTraceRaysKHR.html#VUID-vkCmdTraceRaysKHR-flags-03696
+    // https://registry.khronos.org/vulkan/specs/latest/man/html/vkCmdTraceRaysKHR.html#VUID-vkCmdTraceRaysKHR-flags-03697
+    // https://registry.khronos.org/vulkan/specs/latest/man/html/vkCmdTraceRaysKHR.html#VUID-vkCmdTraceRaysKHR-flags-03512
+    const auto shouldHaveHitGroup = flags & 
+      (PipelineFlags(PipelineFlag::RAY_TRACING_NO_NULL_ANY_HIT_SHADERS_BIT_KHR) | 
+        PipelineFlag::RAY_TRACING_NO_NULL_CLOSEST_HIT_SHADERS_BIT_KHR |
+        PipelineFlag::RAY_TRACING_NO_NULL_INTERSECTION_SHADERS_BIT_KHR);
+    if (shouldHaveHitGroup && !hitGroupsRange.buffer)
+    {
+        NBL_LOG_ERROR("bound pipeline indicates that traceRays command should have hit group, but hitGroupsRange.buffer is null!");
+        return true;
+    }
+
+    // https://registry.khronos.org/vulkan/specs/latest/man/html/vkCmdTraceRaysKHR.html#VUID-vkCmdTraceRaysKHR-flags-03511
+    const auto shouldHaveMissGroup = flags & PipelineFlag::RAY_TRACING_NO_NULL_MISS_SHADERS_BIT_KHR;
+    if (shouldHaveMissGroup && !missGroupsRange.buffer)
+    {
+        NBL_LOG_ERROR("bound pipeline indicates that traceRays command should have hit group, but hitGroupsRange.buffer is null!");
+        return true;
+    }
+
+    const auto& limits = getOriginDevice()->getPhysicalDevice()->getLimits();
+    auto invalidBufferRegion = [this, &limits](const asset::SBufferRange<const IGPUBuffer>& range, uint32_t stride, const char* groupName) -> bool
+    {
+        const IGPUBuffer* const buffer = range.buffer.get();
+
+        if (!buffer) return false;
+
+        if (!range.isValid())
+        {
+            NBL_LOG_ERROR("%s buffer range is not valid!", groupName);
+            return true;
+        }
+
+        if (!(buffer->getCreationParams().usage & IGPUBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT))
+        {
+            NBL_LOG_ERROR("%s buffer must have EUF_SHADER_DEVICE_ADDRESS_BIT usage!", groupName);
+            return true;
+        }
+
+        // https://registry.khronos.org/vulkan/specs/latest/man/html/vkCmdTraceRaysKHR.html#VUID-vkCmdTraceRaysKHR-pRayGenShaderBindingTable-03689
+        if (range.offset % limits.shaderGroupBaseAlignment != 0)
```

Comment:
slightly wrong, you actually want to test `(range.bufer->getDeviceAddress()+offset) % alignment`